### PR TITLE
Add BatchNorm forward folding

### PR DESCRIPTION
**Pull request by elad-c**
Also bug fixes:
 * PyTorch BN folding after ConvTranspose2d
 * PyTorch scale equalization test

### CODE REVIEW

File: model_compression_toolkit/core/common/substitutions/batchnorm_folding.py [hunk 1]
```diff
@@ -150,3 +150,137 @@ def substitute(self,
...
+            framework_attr[self.layer_name_str] = conv_node.name + '_bn'
+
+        weights_dict = {kernel_name: kernel,
+                        self.bias_str: bias}
+
+        conv_bn = copy.deepcopy(conv_node)
+        conv_bn_name = conv_node.name + '_bn'
+        conv_bn.name = conv_bn_name
+        conv_bn.framework_attr = framework_attr
+        conv_bn.weights = weights_dict
```
Comment (eladc-git) [inline model_compression_toolkit/core/common/substitutions/batchnorm_folding.py:272]: it should be "bn_conv" in the variables and in the names

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_folding.py [hunk 2]
```diff
@@ -41,6 +41,22 @@ def batchnorm_folding_node_matchers():
     return bn_node, source_node
 
 
+def batchnorm_forward_folding_node_matchers():
```
Comment (eladc-git) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_folding.py:44]: return type

File: tests/pytorch_tests/model_tests/feature_models/bn_folding_test.py [hunk 3]
```diff
@@ -13,30 +13,102 @@
...
+        set_model(quant_model)
+        out_float = torch_tensor_to_numpy(float_model(*input_x))
+        out_quant = torch_tensor_to_numpy(quant_model(*input_x))
+
+        is_bn_in_model = torch.nn.BatchNorm2d in [type(module) for name, module in quant_model.named_modules()]
+        self.unit_test.assertTrue(self.fold_applied is not is_bn_in_model)
+        self.unit_test.assertTrue(np.isclose(out_quant, out_float, atol=1e-6, rtol=1e-4).all())
+
+
+class BNForwardFoldingNet(torch.nn.Module):
```
Comment (eladc-git) [inline tests/pytorch_tests/model_tests/feature_models/bn_folding_test.py:70]: I would add a test with 2 bn, something like: **bn->conv->bn->relu** to check that it converted to **conv->relu**

**Review by eladc-git — CHANGES_REQUESTED** [review]

File: model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_folding.py [hunk 4]
```diff
@@ -68,6 +84,63 @@ def update_kernel_for_bn_folding_fn(conv_node: BaseNode,
...
+    if conv_node.type == DepthwiseConv2D:
+        kernel_name = DEPTHWISE_KERNEL
+    else:
+        kernel_name = KERNEL
+
+    return kernel, bias + bias_update.flatten(), kernel_name
+
+
+def get_kernel_hw_fn(kernel: np.ndarray):
+    """
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/keras/graph_substitutions/substitutions/batchnorm_folding.py:122]: Shouldn't the name of the function be "get_kernel_fw_shape_fn"?

File: model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_folding.py [hunk 5]
```diff
@@ -48,7 +63,63 @@ def update_kernel_for_bn_folding_fn(conv_node: BaseNode,
...
+    elif conv_node.type == ConvTranspose2d:
+        bias_update = (kernel * bias_factor[:, None, None, None]).sum(axis=0).flatten()
+        _scale = weights_scale[:, None, None, None]
+    else:
+        bias_update = (kernel * bias_factor[None, :, None, None]).sum(axis=1).flatten()
+        _scale = weights_scale[None, :, None, None]
+    return kernel * _scale, bias + bias_update, KERNEL
+
+
+def get_kernel_hw_fn(kernel: np.ndarray):
```
Comment (lior-dikstein) [inline model_compression_toolkit/core/pytorch/graph_substitutions/substitutions/batchnorm_folding.py:101]: same comment about the name as the one in Keras

File: tests/keras_tests/feature_networks_tests/feature_networks/bn_folding_test.py [hunk 6]
```diff
@@ -221,3 +221,43 @@ def create_networks(self):
...
+    def compare(self, quantized_model, float_model, input_x=None, quantization_info=None):
+        is_bn_in_model = any([isinstance(l, tf.keras.layers.BatchNormalization) for l in quantized_model.layers])
+        self.unit_test.assertTrue(self.conversion_applied is not is_bn_in_model)
+        out_float = float_model(input_x)
+        out_quant = quantized_model(input_x)
+        self.unit_test.assertTrue(np.isclose(out_quant, out_float, rtol=1e-4).all())
+
+
+class BNForwardFoldingTest(BaseForwardBatchNormalizationFolding):
+    def __init__(self, unit_test, test_layer, conversion_applied):
```
Comment (lior-dikstein) [inline tests/keras_tests/feature_networks_tests/feature_networks/bn_folding_test.py:248]: Please add a description of what you are testing

File: tests/keras_tests/feature_networks_tests/feature_networks/bn_folding_test.py [hunk 7]
```diff
@@ -221,3 +221,43 @@ def create_networks(self):
...
+        super().__init__(unit_test, conversion_applied)
+        self.test_layer = test_layer
```
Comment (lior-dikstein) [inline tests/keras_tests/feature_networks_tests/feature_networks/bn_folding_test.py:250]: what about testing cases where the BN shouldn't be folded?

File: tests/pytorch_tests/model_tests/feature_models/bn_folding_test.py [hunk 8]
```diff
@@ -13,30 +13,102 @@
...
+    def __init__(self, test_layer):
+        super(BNForwardFoldingNet, self).__init__()
+        self.bn = torch.nn.BatchNorm2d(3)
+        torch.nn.init.uniform_(self.bn.weight, 0.02, 1.05)
+        torch.nn.init.uniform_(self.bn.bias, -1.2, 1.05)
+        torch.nn.init.uniform_(self.bn.running_var, 0.02, 1.05)
+        torch.nn.init.uniform_(self.bn.running_mean, -1.2, 1.05)
+        self.conv = test_layer
+
```
Comment (lior-dikstein) [inline tests/pytorch_tests/model_tests/feature_models/bn_folding_test.py:79]: Please also add descriptions of what you are testing in Pytorch

File: tests/pytorch_tests/model_tests/test_feature_models_runner.py [hunk 9]
```diff
@@ -129,9 +129,25 @@ def test_add_same(self):
 
     def test_bn_folding(self):
         """
-        This test checks the BatchNorm folding feature, plus adding a residual connection.
+        This test checks the BatchNorm folding feature.
+        """
+        BNFoldingNetTest(self, nn.Conv2d(3, 2, kernel_size=1)).run_test()
+        BNFoldingNetTest(self, nn.Conv2d(3, 3, kernel_size=3, groups=3)).run_test()
```
Comment (lior-dikstein) [inline tests/pytorch_tests/model_tests/test_feature_models_runner.py:135]: Can you add comments on the Conv types - for instance, groups=3 is depthwise

**Review by eladc-git — APPROVED** [review]
